2024/articles/omp-santa.pod

```diff
@@ -0,0 +1,339 @@
+    return sqrt(pow(x2 - x1, 2) + pow(y2 - y1, 2) + pow(z2 - z1, 2));
+  }
+  
+  /* C function parallelized with OpenMP */
+  
+  // Main function to calculate the distances from the sleigh to each satellite
+  AV* calculate_distances(double sleigh_lat, double sleigh_lon, double sleigh_alt, AV *satellites) {
+      int satellite_count = av_len(satellites) + 1;  // The number of satellites
+      AV* distances = newAV();  // Create a new Perl array (AV) to hold the distances
+  
+      double sleigh_x, sleigh_y, sleigh_z;
+  
+      // Convert sleigh's geographic coordinates to Cartesian coordinates
+      geo_to_cartesian(sleigh_lat, sleigh_lon, sleigh_alt, &sleigh_x, &sleigh_y, &sleigh_z);
+  
+      // Fetch satellite data into local arrays
+      double *sat_latitudes = malloc(satellite_count * sizeof(double));
+      double *sat_longitudes = malloc(satellite_count * sizeof(double));
+      double *sat_altitudes = malloc(satellite_count * sizeof(double));
+  
+      // Populate satellite data into local arrays (from the Perl array)
+      for (int i = 0; i < satellite_count; i++) {
+        AV *satellite = (AV*) SvRV(*av_fetch(satellites, i, 0));  // Fetch the satellite at index i
+        sat_latitudes[i] = ((double) SvNV(*av_fetch(satellite, 0, 0)));  // Latitude
+        sat_longitudes[i] = ((double) SvNV(*av_fetch(satellite, 1, 0)));  // Longitude
+        sat_altitudes[i] = ((double) SvNV(*av_fetch(satellite, 2, 0)));  // Altitude
+      }
+  
+      // Declare a temporary array to hold distances for each thread
+      double *_distances = malloc(satellite_count * sizeof(double));
+  
+      PerlOMP_GETENV_BASIC // read common environmental variables, provided by OpenMP::Simple 
+  
+      // Start parallel region
+      #pragma omp parallel shared(_distances)
+      {
+        // Parallel for loop to compute distances in parallel
+        #pragma omp for
+        for (int i = 0; i < satellite_count; i++) {
+            _distances[i] = 0.0;  // Initialize
+  
+            double sat_x, sat_y, sat_z;
+  
+            // Convert satellite's geographic coordinates to Cartesian coordinates
+            geo_to_cartesian(sat_latitudes[i], sat_longitudes[i], sat_altitudes[i], &sat_x, &sat_y, &sat_z);
+  
+            // Calculate the distance from the sleigh to this satellite
+            _distances[i] = calculate_distance(sleigh_x, sleigh_y, sleigh_z, sat_x, sat_y, sat_z);
+        }
+      }
+  
+      // Combine the results from all threads into the main distances array inside the parallel region
+      for (int i = 0; i < satellite_count; i++) {
+        av_push(distances, newSVnv(_distances[i]));
+      }
+  
+      // Free the private distance arrays and satellite data arrays
+      free(_distances);
+      free(sat_latitudes);
+      free(sat_longitudes);
+      free(sat_altitudes);
+  
+      // Return the AV containing the distances to Perl
+      return distances;
+  }
+  
+  __END__
+
+=end perl
```

Comment:
Is there some value added to the article by actually including it here? Maybe just point to a repo or a gist?